File: week6/credit/credit.py
from cs50 import get_string


def main():
    while(True):
        number = get_string("Number: ")
        if int(number) > 0:
            break

    # Checksum
    # Multiply every other digit by 2, starting with second-last digit
    # Add those products' digits together
    # Add the sum to the sum of the digits that weren't multiplied by 2
    # if the total's last digit is 0, number is valid!
    first_set = 0
    second_set = 0

    for i in range(len(number), 0, -1):
        if (i - len(number)) % 2 != 0:
            other = int(number[i - 1]) * 2
            if other > 9:
                first_set += other % 10 + other // 10
            else:
                first_set += other
        else:
            second_set += int(number[i - 1])

    total = str(first_set + second_set)

    if total[-1] == "0":
        if number[0] == "4" and (len(number) == 13 or len(number) == 16):
            print("VISA")
        elif (number[0:2] >= "51" and number[0:2] <= "55") and len(number) == 16:
            print("MASTERCARD")
        elif (number[0:2] == "34" or number[0:2] == "37") and len(number) == 15:
            print("AMEX")
        else:
            print("INVALID")
    else:
        print("INVALID")


main()

File: week1/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    long number;

    do
    {
        number = get_long("Number: ");
    }
    while (!number);

    int counter = 0;
    int first_set = 0;
    int second_set = 0;
    int remainder = 0;
    int start_number = 0;

    while (number > 0)
    {
        counter ++;
        if (number > 9)
        {
            start_number = number;
        }

        remainder = number % 10;

        if (counter % 2 == 0)
        {
            remainder *= 2;
            if (remainder > 9)
            {
                remainder = remainder % 10 + remainder / 10;
            }

            first_set += remainder;
        }
        else
        {
            second_set += remainder;
        }
        number = number / 10;
    }
    int sum = first_set + second_set ;

    if (sum % 10 == 0)
    {
        if (start_number / 10 == 4 && (counter == 13 || counter == 16))
        {
            printf("VISA\n");
        }
        else if ((start_number >= 51 && start_number <= 55)  && counter == 16)
        {
            printf("MASTERCARD\n");
        }
        else if ((start_number == 34 || start_number == 37) && counter == 15)
        {
            printf("AMEX\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
    else
    {
        printf("INVALID\n");
    }
}

